File: ios/Models/Parlor.swift
import Foundation

/**
 * FED-SPA data model - mirrors data/meta/schema.json definitions.
 *
 * LicensedParlor  <- data/public/licensed.json ("parlors" array)
 * UnlicensedParlor <- decrypted subscriber envelope ("parlors" array)
 */
struct LicensedParlor: Codable, Identifiable, Hashable {
    let licenseNumber: String
    let businessName: String
    let profession: String
    let status: String
    let expirationDate: String
    let originalIssueDate: String
    let address: Address
    let county: String
    let disciplineOnFile: Bool
    let publicComplaint: Bool
    let dataAsOf: String
    let lastChecked: String
    var notes: String?

    var id: String { licenseNumber }

    enum CodingKeys: String, CodingKey {
        case licenseNumber = "license_number"
        case businessName = "business_name"
        case profession
        case status
        case expirationDate = "expiration_date"
        case originalIssueDate = "original_issue_date"
        case address
        case county
        case disciplineOnFile = "discipline_on_file"
        case publicComplaint = "public_complaint"
        case dataAsOf = "data_as_of"
        case lastChecked = "last_checked"
        case notes
    }
}

struct UnlicensedParlor: Codable, Identifiable, Hashable {
    let businessName: String
    let address: Address
    let status: String
    let lastChecked: String
    let reason: String
    let county: String

    var id: String { "\(businessName)-\(address.street)" }

    enum CodingKeys: String, CodingKey {
        case businessName = "business_name"
        case address
        case status
        case lastChecked = "last_checked"
        case reason
        case county
    }
}

struct Address: Codable, Hashable {
    let street: String
    let city: String
    let state: String
    let zip: String
}

/** Top-level document for the public tier. */
struct LicensedDocument: Codable {
    let version: Int
    let asOf: String
    let source: String
    let sourceUrl: String
    let updateFrequency: String
    let parlors: [LicensedParlor]

    enum CodingKeys: String, CodingKey {
        case version
        case asOf = "as_of"
        case source
        case sourceUrl = "source_url"
        case updateFrequency = "update_frequency"
        case parlors
    }
}

/** Top-level document for the subscriber tier (post-decryption). */
struct UnlicensedDocument: Codable {
    let version: Int
    let asOf: String
    let parlors: [UnlicensedParlor]

    enum CodingKeys: String, CodingKey {
        case version
        case asOf = "as_of"
        case parlors
    }
}

File: android/app/src/main/java/ai/ninjatech/fedspa/Parlor.kt
package ai.ninjatech.fedspa

/**
 * One establishment row, used for BOTH tiers.
 *
 * Licensed rows come from assets/licensed.json (public, shipped in every
 * build). Unlicensed rows come from the decrypted subscriber envelope -
 * same shape, minus the license fields (licenseNumber is "—").
 *
 * Field names map 1:1 to data/meta/schema.json definitions:
 *   licensed_parlor / unlicensed_parlor.
 */
data class Parlor(
    val licenseNumber: String,
    val businessName: String,
    val profession: String,
    val status: String,
    val expirationDate: String,
    val originalIssueDate: String,
    val street: String,
    val city: String,
    val state: String,
    val zip: String,
    val county: String,
    val disciplineOnFile: Boolean,
    val publicComplaint: Boolean,
    val dataAsOf: String,
    val lastChecked: String,
    val notes: String?
) {
    /** "Boynton Beach, FL" style label for list rows. */
    val cityLine: String get() = "$city, $state"

    /** True when this row is from the subscriber watchlist tier. */
    val isUnlicensed: Boolean get() = licenseNumber == "—"

    /** Full mailing address on one line. */
    val addressLine: String get() = "$street, $city, $state $zip"

    /**
     * Days until expirationDate (ISO yyyy-mm-dd), or null when the row
     * has no expiration (unlicensed tier) or the date is unparseable.
     * Negative = already expired.
     */
    fun daysUntilExpiry(): Int? {
        if (expirationDate.isEmpty()) return null
        return try {
            val exp = java.time.LocalDate.parse(expirationDate)
            java.time.LocalDate.now().until(exp, java.time.temporal.ChronoUnit.DAYS).toInt()
        } catch (e: Exception) {
            null
        }
    }

    /** The web UI shows an "expiring soon" chip at <= 180 days; match it. */
    val expiringSoon: Boolean get() = (daysUntilExpiry() ?: 999) in 0..180
}
